diff --git a/src/nfa/mcclellancompile.cpp b/src/nfa/mcclellancompile.cpp
index c1a4f87..a8cbb15 100644
--- a/src/nfa/mcclellancompile.cpp
+++ b/src/nfa/mcclellancompile.cpp
@@ -1400,6 +1400,19 @@ void generate_symbol_chain(dfa_info &info, vector<symbol_t> &chain_tail) {
     }
 }

+template<class _BidIt> inline
+       void reverseit(_BidIt _First, _BidIt _Last)
+       {       // reverse elements in [_First, _Last)
+       _Reverseit(_First, _Last);
+       }
+
+template<class _BidIt> inline
+       void _Reverseit(_BidIt _First, _BidIt _Last)
+       {       // reverse elements in [_First, _Last), bidirectional iterators
+       for (; _First != _Last && _First != --_Last; ++_First)
+               iter_swap(_First, _Last);
+       }
+
 /* \brief Find potential regions of states to be packed into wide states. */
 static
 void find_wide_state(dfa_info &info) {
@@ -1444,7 +1457,7 @@ void find_wide_state(dfa_info &info) {
                     DEBUG_PRINTF("Newly pushed state: s%u\n", head);
                 }

-                reverse(temp_chain.begin(), temp_chain.end());
+                reverseit(temp_chain.begin(), temp_chain.end());
                 temp_chain.push_back(curr_id);

                 assert(head > 0 && head == temp_chain.front());
